README.md


# MaAsLin2 User Manual #

MaAsLin2 is the next generation of MaAsLin (Microbiome Multivariable Association with Linear Models).

[MaAsLin2](http://huttenhower.sph.harvard.edu/maaslin2) is comprehensive R package for efficiently determining multivariable association between clinical metadata and microbial meta-omics features. MaAsLin2 relies on general linear models to accommodate most modern epidemiological study designs, including cross-sectional and longitudinal, along with a variety of filtering, normalization, and transform methods.

If you use the MaAsLin2 software, please cite our manuscript: 

Mallick H, Rahnavard A, McIver LJ, Ma S, Zhang Y, Nguyen LH, Tickle TL, Weingart G, Ren B, Schwager EH, Chatterjee S, Thompson KN, Wilkinson JE, Subramanian A, Lu Y, Waldron L, Paulson JN, Franzosa EA, Bravo HC, Huttenhower C (2021). [Multivariable Association Discovery in Population-scale Meta-omics Studies](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009442). PLoS Computational Biology, 17(11):e1009442.

Check out the [MaAsLin 2 tutorial](https://github.com/biobakery/biobakery/wiki/maaslin2) for an overview of analysis options.

If you have questions, please direct it to :   
[MaAsLin2 Forum](https://forum.biobakery.org/c/Downstream-analysis-and-statistics/MaAsLin2)    
[Google Groups](https://groups.google.com/forum/#!forum/maaslin-users) (Read only)

![](https://github.com/biobakery/Maaslin2/workflows/build%20and%20test/badge.svg)

<a href="http://www.bioconductor.org/packages/devel/bioc/html/Maaslin2.html#since"><img border="0" src="http://www.bioconductor.org/shields/years-in-bioc/Maaslin2.svg" title="How long since the package was first in a released Bioconductor version (or is it in devel only)."></a> 

--------------------------------------------

## Contents ##
* [Description](#description)
* [Requirements](#requirements)
* [Installation](#installation)
* [How to Run](#how-to-run)
    * [Input Files](#input-files)
    * [Output Files](#output-files)
    * [Run a Demo](#run-a-demo)
    * [Options](#options)
* [Visualization](#visualization)
* [Troubleshooting](#troubleshooting)

## Description ##

MaAsLin2 finds associations between microbiome multi-omics features and complex metadata in population-scale epidemiological studies. The software includes multiple analysis methods (including support for multiple covariates and repeated measures), filtering, normalization, and transform options to customize analysis for your specific study. 

## Requirements ##

MaAsLin2 is an R package that can be run on the command line or as an R function.

## Installation ##

MaAsLin2 can be run from the command line or as an R function.

If only running from the command line, you do not need to install the MaAsLin2 package but you will need to install the MaAsLin2 dependencies.

### From command line ###

1. Download the source: [MaAsLin2.master.zip](https://github.com/biobakery/Maaslin2/archive/master.zip)
2. Decompress the download: 
    * ``$ tar xzvf Maaslin2-master.zip``
3. Install the Bioconductor dependencies edgeR and metagenomeSeq. 
4. Install the CRAN dependencies:
    * ``$ R -q -e "install.packages(c('lmerTest','pbapply','car','dplyr','vegan','chemometrics','ggplot2','pheatmap','hash','logging','data.table','glmmTMB','MASS','cplm','pscl'), repos='http://cran.r-project.org')"``
5. Install the MaAsLin2 package (only required if running as an R function): 
    * ``$ R CMD INSTALL maaslin2``

### From R ###

To install the latest release version of MaAsLin 2:

```{r, eval=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("Maaslin2")
```

To install the latest development version of MaAsLin 2:

```{r, eval=FALSE}
install.packages("devtools")
library("devtools")
install_github("biobakery/Maaslin2")
```

## How to Run ##

MaAsLin2 can be run from the command line or as an R function. Both 
methods require the same arguments, have the same options, 
and use the same default settings.

### Input Files ###

MaAsLin2 requires two input files.

1. Data (or features) file
    * This file is tab-delimited.
    * Formatted with features as columns and samples as rows.
    * The transpose of this format is also okay.
    * Possible features in this file include taxonomy or genes.
2. Metadata file
    * This file is tab-delimited.
    * Formatted with features as columns and samples as rows.
    * The transpose of this format is also okay.
    * Possible metadata in this file include gender or age.

The data file can contain samples not included in the metadata file
(along with the reverse case). For both cases, those samples not 
included in both files will be removed from the analysis. 
Also the samples do not need to be in the same order in the two files.

NOTE: If running MaAsLin2 as a function, the data and metadata 
inputs can be of type ``data.frame`` instead of a path to a file.

### Output Files ###

MaAsLin2 generates two types of output files: data and visualization.

1. Data output files
    * ``all_results.tsv``
        * This includes the same data as the data.frame returned.
        * This file contains all results ordered by increasing q-value.
        * The first columns are the metadata and feature names.
        * The next two columns are the value and coefficient from the model.
        * The next column is the standard deviation from the model.
        * The ``N`` column is the total number of data points.
        * The ``N.not.zero`` column is the total of non-zero data points.
        * The pvalue from the calculation is the second to last column.
        * The qvalue is computed with `p.adjust` with the correction method.
    * ``significant_results.tsv``
        * This file is a subset of the results in the first file.
        * It only includes associations with q-values <= to the threshold.
    * ``features```
        * This folder includes the filtered, normalized, and transformed versions of the input feature table.
        * These steps are performed sequentially in the above order.
        * If an option is set such that a step does not change the data, the resulting table will still be output.
    * ``models.rds``
        * This file contains a list with every model fit object.
        * It will only be generated if save_models is set to TRUE.
    * ``residuals.rds``
        * This file contains a data frame with residuals for each feature.
    * ``fitted.rds``
        * This file contains a data frame with fitted values for each feature.
    * ``ranef.rds``
        * This file contains a data frame with extracted random effects for each feature (if random effects are specified).
    * ``maaslin2.log``
        * This file contains all log information for the run.
        * It includes all settings, warnings, errors, and steps run.
2. Visualization output files
    * ``heatmap.pdf``
        * This file contains a heatmap of the significant associations.
    * ``[a-z/0-9]+.pdf``
        * A plot is generated for each significant association.
        * Scatter plots are used for continuous metadata.
        * Box plots are for categorical data.
        * Data points plotted are after filtering but prior to normalization and transform.

### Run a Demo ###

Example input files can be found in the ``inst/extdata`` folder 
of the MaAsLin2 source. The files provided were generated from
the HMP2 data which can be downloaded from https://ibdmdb.org/ .

``HMP2_taxonomy.tsv``: is a tab-demilited file with species as columns and samples as rows. It is a subset of the taxonomy file so it just includes the species abundances for all samples.

``HMP2_metadata.tsv``: is a tab-delimited file with samples as rows and metadata as columns. It is a subset of the metadata file so that it just includes some of the fields.


#### Command line ####

``$ Maaslin2.R --fixed_effects="diagnosis,dysbiosisnonIBD,dysbiosisUC,dysbiosisCD,antibiotics,age" --random_effects="site,subject" --standardize=FALSE inst/extdata/HMP2_taxonomy.tsv inst/extdata/HMP2_metadata.tsv demo_output``

* Make sure to provide the full path to the MaAsLin2 executable (ie ./R/Maaslin2.R).
* In the demo command:
    * ``HMP2_taxonomy.tsv`` is the path to your data (or features) file
    * ``HMP2_metadata.tsv`` is the path to your metadata file
    * ``demo_output`` is the path to the folder to write the output


#### In R ####

```{r}
library(Maaslin2)
input_data <- system.file(
    'extdata','HMP2_taxonomy.tsv', package="Maaslin2")
input_metadata <-system.file(
    'extdata','HMP2_metadata.tsv', package="Maaslin2")
fit_data <- Maaslin2(
    input_data, input_metadata, 'demo_output'
    fixed_effects = c('diagnosis', 'dysbiosisnonIBD','dysbiosisUC','dysbiosisCD', 'antibiotics', 'age'),
    random_effects = c('site', 'subject'),
    standardize = FALSE)
```

##### Session Info #####

Session info from running the demo in R can be displayed with the following command.

```{r}
sessionInfo()
```

### Options ###

Run MaAsLin2 help to print a list of the options and the default settings.


$ Maaslin2.R --help
Usage: ./R/Maaslin2.R [options] <data.tsv> <metadata.tsv> <output_folder>


Options:
    -h, --help
        Show this help message and exit

    -a MIN_ABUNDANCE, --min_abundance=MIN_ABUNDANCE
        The minimum abundance for each feature [ Default: 0 ]   

    -p MIN_PREVALENCE, --min_prevalence=MIN_PREVALENCE
        The minimum percent of samples for which a feature 
        is detected at minimum abundance [ Default: 0.1 ]

    -b MIN_VARIANCE, --min_variance=MIN_VARIANCE
       Keep features with variance greater than [ Default: 0.0 ]

    -s MAX_SIGNIFICANCE, --max_significance=MAX_SIGNIFICANCE
        The q-value threshold for significance [ Default: 0.25 ]

    -n NORMALIZATION, --normalization=NORMALIZATION
        The normalization method to apply [ Default: TSS ]
        [ Choices: TSS, CLR, CSS, NONE, TMM ]

    -t TRANSFORM, --transform=TRANSFORM
        The transform to apply [ Default: LOG ]
        [ Choices: LOG, LOGIT, AST, NONE ]

    -m ANALYSIS_METHOD, --analysis_method=ANALYSIS_METHOD
        The analysis method to apply [ Default: LM ]
        [ Choices: LM, CPLM, NEGBIN, ZINB ]

    -r RANDOM_EFFECTS, --random_effects=RANDOM_EFFECTS
        The random effects for the model, comma-delimited
        for multiple effects [ Default: none ]

    -f FIXED_EFFECTS, --fixed_effects=FIXED_EFFECTS
        The fixed effects for the model, comma-delimited
        for multiple effects [ Default: all ]

    -c CORRECTION, --correction=CORRECTION
        The correction method for computing the 
        q-value [ Default: BH ]

    -z STANDARDIZE, --standardize=STANDARDIZE
        Apply z-score so continuous metadata are 
        on the same scale [ Default: TRUE ]

    -l PLOT_HEATMAP, --plot_heatmap=PLOT_HEATMAP
        Generate a heatmap for the significant 
        associations [ Default: TRUE ]

    -i HEATMAP_FIRST_N, --heatmap_first_n=HEATMAP_FIRST_N
        In heatmap, plot top N features with significant 
        associations [ Default: TRUE ]

    -o PLOT_SCATTER, --plot_scatter=PLOT_SCATTER
        Generate scatter plots for the significant
        associations [ Default: TRUE ]
        
    -g MAX_PNGS, --max_pngs=MAX_PNGS
        The maximum number of scatter plots for signficant associations 
        to save as png files [ Default: 10 ]
    
    -O SAVE_SCATTER, --save_scatter=SAVE_SCATTER
        Save all scatter plot ggplot objects
        to an RData file [ Default: FALSE ]

    -e CORES, --cores=CORES
        The number of R processes to run in parallel
        [ Default: 1 ]
        
    -j SAVE_MODELS --save_models=SAVE_MODELS
        Return the full model outputs and save to an RData file
        [ Default: FALSE ]
    
    -d REFERENCE, --reference=REFERENCE
        The factor to use as a reference level for a categorical variable 
        provided as a string of 'variable,reference', semi-colon delimited for 
        multiple variables. Not required if metadata is passed as a factor or 
        for variables with less than two levels but can be set regardless.
        [ Default: NA ] 

## Contributions ##
Thanks go to these wonderful people:

- Nick Waters <nickp60@gmail.com>
    * Design of the PR and attribution process

## Troubleshooting ##

1. Question: When I run from the command line I see the error ``Maaslin2.R: command not found``. How do I fix this? 
    * Answer: Provide the full path to the executable when running Maaslin2.R.
2. Question: When I run as a function I see the error ``Error in library(Maaslin2): there is no package called 'Maaslin2'``. How do I fix this? 
    * Answer: Install the R package and then try loading the library again.
3. Question: When I try to install the R package I see errors about dependencies not being installed. Why is this?
    * Answer: Installing the R package will not automatically install the packages MaAsLin2 requires. Please install the dependencies and then install the MaAsLin2 R package.



NEWS
Changes in version 1.7.4 (05-09-2023):
* Resolve "cannot xtfrm data frames" error
* Add argument to save full model outputs
* Add argument to set number of pngs saved
* Update reference level handling to accept factors
* Allow for variable to be a fixed and random effect
* Allow tibble input
* Remove unused dependencies
* Update build check to current bioconductor standards

Changes in version 1.7.3 (11-02-2021):
* Update sourcing R scripts 
* Update reference level handling
* Update example files
* Fix saved plots to allow for runs with less then N plots per metadata
* Convert strings to numeric when needed for metadata
* Fix actions to build with latest version of dependencies

Changes in version 1.7.1 (07-27-2021):
* Update tutorial data files
* Update knitr dependency for bioconductor tests

Changes in version 1.5.1 (12-02-2020):
* Update log from base 10 to base 2.
* ZICP is now deprecated (https://cran.r-project.org/web/packages/cplm/NEWS)
* SLM is removed in favor of a future R2 functionality for all models
* Fitted values are returned along with residuals
* Extracted random effects are also returned

Changes in version 1.3.2 (08-12-2020):
* Resolve issue with x labels missing from boxplots for metadata variables without levels and increase max jpgs written. 
* Update the check for variables with more than two levels that will require a reference provided by the user for the model and the boxplots to be more strict (ignore UNK in the level count, don't check if it is a factor, and check to see if all of the levels are numeric to ignore continuous variables).
* Fix ZINB error.

Changes in version 1.3.1 (06-16-2020):
+ Add random effects capability to all the other non-LM models.
+ Add variance filtering option with default set to 0.
+ Normalization is now performed after filtering and N.not.zero is calculated on the untransformed space. Also includes minor edits to synchronize with the manuscript.
+ Add reference option required for fixed effects variables with more than two levels. Reference used in model and for boxplots.
+ Update heatmap to include all categorical levels.

Changes in version 1.1.2 (04-17-2020):
+ Small changes to pass build tests for latest bioconductor release.

Changes in version 1.1.1 (01-20-2020):
+ Updated reference to sync up with manuscript.

Changes in version 0.99.19 (2019-09-27):
+ Add figures folder of jpgs including heatmap and most significant scatter/box plots per metadata to be used in auto-generated workflow reports.

Changes in version 0.99.18 (2019-09-27):
+ For log function if x is zero use half of min value instead of one (merged changed function from Tommi Vatanen's fork).

Changes in version 0.99.17 (2019-08-20):
+ Filtering by min abundance is limited to greater than (prior was greater than and equal to)

Changes in version 0.99.16 (2019-08-05):
+ Add value to categorical plots.

Changes in version 0.99.15 (2019-07-31):
+ Adding another dependency required by bioconductor MacOS automated build/test

Changes in version 0.99.14 (2019-07-31):
+ Adding two more dependencies required by bioconductor MacOS automated build/test

Changes in version 0.99.13 (2019-07-31):
+ Iterations for addition to bioconductor: Add back in license file and small changes to coding sections of vignette.
+ Modifications to man page to include new option.

Changes in version 0.99.12 (2019-07-26):
+ Add new option to set the max number of features shown in heatmap.
+ Fix heatmap to include all rows of significant values for the top N features instead of only including the rows after finding the top N features.

Changes in version 0.99.11 (2019-07-24):
+ Update demo to data from HMP2 (provided by Himel).

Changes in version 0.99.10 (2019-07-19):
+ Small modifications to documentation to update dependency install notes to match bioconductor.

Changes in version 0.99.9 (2019-07-18):
+ Add dependency to namespace for automated tests.

Changes in version 0.99.8 (2019-07-18):
+ Add one more dependency for bioconductor MacOS build tests

Changes in version 0.99.7 (2019-07-17):
+ Change file paths to relative to package for windows tests.

Changes in version 0.99.6 (2019-07-17):
+ Fix format of R sections in vignette to pass tests.

Changes in version 0.99.5 (2019-07-17):
+ Update R sections of vignette format.
+ Change test paths for windows.

Changes in version 0.99.4 (2019-07-17):
+ Update required R version.
+ Modifications in vignette format for bioconductor build.

Changes in version 0.99.3 (2019-07-17):
+ Modifications to package based on feedback from bioconductor review (additions to description, notes on data files, update testing to testthat, condensing vignette sections and updating install to bioconductor method, remove tests for packages, use seq_len and lapply)

Changes in version 0.99.2 (2019-06-27):
+ Show top N features in heatmap instead of top N associations

Changes in version 0.99.1 (2019-06-05):
+ Fix plots to allow for NAs in values

Changes in version 0.99.0 (2019-05-24):
+ Only show the top 50 associations in the heatmaps
+ Use static heatmap plot colors
+ In boxplots, use angle for x axis text for long text strings (if any in set is over 5 chars)
+ For larger y axis labels reduce the font size (if over 15 chars)
+ Add Ns to plots in annotation for continuous and x axis label for categorical

Changes in version 0.3.0 (2019-05-20):
+ Plots now show normalized/filtered/transformed data
+ Package modifications for submission to bioconductor

Changes in version 0.2.3 (2018-12-20):
+ Move filtering to after normalization
+ Updates to barplots

Changes in version 0.2.2 (2018-11-15):
+ Fix issue with single column in visualizations (Thanks, sma!)
+ Add hash to dependencies
+ Change output column names to match data.frame names
+ Add options to bypass plotting
+ Add crossed random effects for LM using lme4 and lmerTest
+ Fix ZICP fitting errors
+ Add stderr to results
+ Rotate heatmap column names by 45 degrees

Changes in version 0.2.1 (2018-10-10):

R/viz.R
# MaAsLin2 theme based on Nature journal requirements
nature_theme <- function(x_axis_labels, y_label) {
  # set default text format based on categorical and length
  angle <- NULL
  hjust <- NULL
  size <- 8
  if (max(nchar(x_axis_labels), na.rm = TRUE) > 5) {
    angle <- 45
    hjust <- 1
    size <- 6
  }
  axis_title_size <- 10
  if (nchar(y_label) > 15) {
    axis_title_size <- 8
  }
  if (nchar(y_label) > 25) {
    axis_title_size <- 6
  }
  return(ggplot2::theme_bw() + ggplot2::theme(
    axis.text.x = ggplot2::element_text(size = size, vjust = 1, hjust = hjust, angle = angle),
    axis.text.y = ggplot2::element_text(size = 8, hjust = 1),
    axis.title = ggplot2::element_text(size = axis_title_size),
    plot.title = ggplot2::element_text(size = 7, face = "bold"),
    legend.title = ggplot2::element_text(size = 6, face = "bold"),
    legend.text = ggplot2::element_text(size = 6),
    axis.line = ggplot2::element_line(colour = "black", size = .25),
    axis.line.x = ggplot2::element_line(colour = "black", size = .25),
    axis.line.y = ggplot2::element_line(colour = "black", size = .25),
    panel.border = ggplot2::element_blank(),
    panel.grid.major = ggplot2::element_blank(),
    panel.grid.minor = ggplot2::element_blank()
  ))
}


# MaAsLin2 heatmap function for overall view of associations
maaslin2_heatmap <-
  function(output_results,
           title = NA,
           cell_value = "qval",
           data_label = "data",
           metadata_label = "metadata",
           border_color = "grey93",
           color = colorRampPalette(c("darkblue", "grey90", "darkred")),
           col_rotate = 90,
           first_n = 50) {
    # read MaAsLin output
    df <- read.table(
      output_results,
      header = TRUE,
      sep = "\t",
      fill = TRUE,
      comment.char = "",
      check.names = FALSE
    )

    title_additional <- ""

    title_additional <- ""
    if (!is.na(first_n) & first_n > 0 & first_n < dim(df)[1]) {
      if (cell_value == "coef") {
        df <- df[order(-abs(df[[cell_value]])), ]
      } else {
        df <- df[order(df[[cell_value]]), ]
      }
      # get the top n features with significant associations
      df_sub <- df[1:first_n, ]
      for (first_n_index in seq(first_n, dim(df)[1]))
      {
        if (length(unique(df_sub$feature)) == first_n) {
          break
        }
        df_sub <- df[1:first_n_index, ]
      }
      # get all rows that have the top N features
      df <- df[which(df$feature %in% df_sub$feature), ]
      title_additional <- paste("Top", first_n, sep = " ")
    }

    if (dim(df)[1] < 2) {
      print("There are no associations to plot!")
      return(NULL)
    }

    metadata <- df$metadata
    data <- df$feature
    dfvalue <- df$value
    value <- NA

    # values to use for coloring the heatmap
    # and set the colorbar boundaries
    if (cell_value == "pval") {
      value <- -log(df$pval) * sign(df$coef)
      value <- pmax(-20, pmin(20, value))
      if (is.null(title)) {
        title <- "(-log(pval)*sign(coeff))"
      }
    } else if (cell_value == "qval") {
      value <- -log(df$qval) * sign(df$coef)
      value <- pmax(-20, pmin(20, value))
      if (is.null(title)) {
        title <- "(-log(qval)*sign(coeff))"
      }
    } else if (cell_value == "coef") {
      value <- df$coef
      if (is.null(title)) {
        title <- "(coeff)"
      }
    }

    if (title_additional != "") {
      title <- paste(title_additional, "features with significant associations", title, sep = " ")
    } else {
      title <- paste("Significant associations", title, sep = " ")
    }

    # identify variables with more than one level present
    verbose_metadata <- c()
    metadata_multi_level <- c()
    for (i in unique(metadata)) {
      levels <- unique(df$value[df$metadata == i])
      if (length(levels) > 1) {
        metadata_multi_level <- c(metadata_multi_level, i)
        for (j in levels) {
          verbose_metadata <- c(verbose_metadata, paste(i, j))
        }
      } else {
        verbose_metadata <- c(verbose_metadata, i)
      }
    }

    n <- length(unique(data))
    m <- length(unique(verbose_metadata))

    if (n < 2) {
      print(
        paste(
          "There is not enough features in the associations",
          "to create a heatmap plot.",
          "Please review the associations in text output file."
        )
      )
      return(NULL)
    }

    if (m < 2) {
      print(
        paste(
          "There is not enough metadata in the associations",
          "to create a heatmap plot.",
          "Please review the associations in text output file."
        )
      )
      return(NULL)
    }

    a <- matrix(0, nrow = n, ncol = m)
    a <- as.data.frame(a)

    rownames(a) <- unique(data)
    colnames(a) <- unique(verbose_metadata)

    for (i in seq_len(dim(df)[1])) {
      current_metadata <- metadata[i]
      if (current_metadata %in% metadata_multi_level) {
        current_metadata <- paste(metadata[i], dfvalue[i])
      }
      if (abs(a[
        as.character(data[i]),
        as.character(current_metadata)
      ]) > abs(value[i])) {
        next
      }
      a[as.character(data[i]), as.character(current_metadata)] <- value[i]
    }

    # get the range for the colorbar
    max_value <- ceiling(max(a))
    min_value <- ceiling(min(a))
    range_value <- max(c(abs(max_value), abs(min_value)))
    breaks <- seq(-1 * range_value, range_value, by = 1)

    p <- NULL
    tryCatch(
      {
        p <-
          pheatmap::pheatmap(
            a,
            cellwidth = 5,
            cellheight = 5,
            # changed to 3
            main = title,
            fontsize = 6,
            kmeans_k = NA,
            border = TRUE,
            show_rownames = TRUE,
            show_colnames = TRUE,
            scale = "none",
            cluster_rows = FALSE,
            cluster_cols = TRUE,
            clustering_distance_rows = "euclidean",
            clustering_distance_cols = "euclidean",
            legend = TRUE,
            border_color = border_color,
            color = color(range_value * 2),
            breaks = breaks,
            treeheight_row = 0,
            treeheight_col = 0,
            display_numbers = matrix(ifelse(
              a > 0.0, "+", ifelse(a < 0.0, "-", "")
            ), nrow(a)),
            silent = TRUE
          )
      },
      error = function(err) {
        logging::logerror("Unable to plot heatmap")
        logging::logerror(err)
      }
    )
    return(p)
  }

save_heatmap <-
  function(results_file,
           heatmap_file,
           figures_folder,
           title = NULL,
           cell_value = "qval",
           data_label = "data",
           metadata_label = "metadata",
           border_color = "grey93",
           color = colorRampPalette(c("blue", "grey90", "red")),
           first_n = 50) {
    # generate a heatmap and save it to a pdf and as a png
    heatmap <-
      maaslin2_heatmap(
        results_file,
        title,
        cell_value,
        data_label,
        metadata_label,
        border_color,
        color,
        first_n
      )

    if (!is.null(heatmap)) {
      pdf(heatmap_file)
      print(heatmap)
      dev.off()

      png_file <- file.path(figures_folder, "heatmap.png")
      png(png_file, res = 150, height = 800, width = 1100)
      print(heatmap)
      dev.off()
    }
  }

maaslin2_association_plots <-
  function(metadata,
           features,
           output_results,
           write_to = "./",
           figures_folder = "./figures/",
           max_pngs = 10,
           save_scatter = FALSE) {
    # MaAslin2 scatter plot function and theme

    # combine the data and metadata to one datframe using common rows
    # read MaAsLin output
    if (is.character(metadata)) {
      metadata <- read.table(
        metadata,
        header = TRUE,
        row.names = 1,
        sep = "\t",
        fill = FALSE,
        comment.char = "",
        check.names = FALSE
      )
    }
    if (is.character(features)) {
      features <- read.table(
        features,
        header = TRUE,
        row.names = 1,
        sep = "\t",
        fill = FALSE,
        comment.char = "",
        check.names = FALSE
      )
    }

    common_rows <- intersect(rownames(features), rownames(metadata))
    input_df_all <-
      cbind(
        features[common_rows, , drop = FALSE],
        metadata[common_rows, , drop = FALSE]
      )

    # read MaAsLin output
    if (is.character(output_results)) {
      output_df_all <- read.table(
        output_results,
        header = TRUE,
        row.names = NULL,
        sep = "\t",
        fill = FALSE,
        comment.char = "",
        check.names = FALSE
      )
    } else {
      output_df_all <- output_results
    }

    if (dim(output_df_all)[1] < 1) {
      print("There are no associations to plot!")
      return(NULL)
    }

    logging::loginfo(
      paste(
        "Plotting associations from most",
        "to least significant,",
        "grouped by metadata"
      )
    )
    metadata_types <- unlist(output_df_all[, "metadata"])
    metadata_labels <-
      unlist(metadata_types[!duplicated(metadata_types)])
    metadata_number <- 1
    saved_plots <- list()

    for (label in metadata_labels) {
      saved_plots[[label]] <- list()
      # for file name replace any non alphanumeric with underscore
      plot_file <-
        paste(
          write_to,
          "/",
          gsub("[^[:alnum:]_]", "_", label),
          ".pdf",
          sep = ""
        )
      data_index <- which(label == metadata_types)
      logging::loginfo(
        "Plotting data for metadata number %s, %s",
        metadata_number,
        label
      )
      pdf(
        plot_file,
        width = 2.65,
        height = 2.5,
        onefile = TRUE
      )

      x <- NULL
      y <- NULL
      count <- 1
      for (i in data_index) {
        x_label <- as.character(output_df_all[i, "metadata"])
        y_label <- as.character(output_df_all[i, "feature"])
        results_value <- as.character(output_df_all[i, "value"])
        qval <- as.numeric(output_df_all[i, "qval"])
        coef_val <- as.numeric(output_df_all[i, "coef"])
        input_df <- input_df_all[c(x_label, y_label)]
        colnames(input_df) <- c("x", "y")

        # if Metadata is continuous generate a scatter plot
        # Continuous is defined as numerical with more than
        # 2 values (to exclude binary data)
        temp_plot <- NULL
        if (is.numeric(input_df[1, "x"]) &
          length(unique(input_df[["x"]])) > 2) {
          logging::loginfo(
            "Creating scatter plot for continuous data, %s vs %s",
            x_label,
            y_label
          )
          temp_plot <- ggplot2::ggplot(
            data = input_df,
            ggplot2::aes(
              as.numeric(as.character(x)),
              as.numeric(as.character(y))
            )
          ) +
            ggplot2::geom_point(
              fill = "darkolivegreen4",
              color = "black",
              alpha = .5,
              shape = 21,
              size = 1,
              stroke = 0.15,
              na.rm = TRUE
            ) +
            ggplot2::scale_x_continuous(
              limits = c(min(
                input_df["x"]
              ), max(input_df["x"]))
            ) +
            ggplot2::scale_y_continuous(
              limits = c(min(
                input_df["y"]
              ), max(input_df["y"]))
            ) +
            ggplot2::stat_smooth(
              formula = y ~ x,
              method = "glm",
              size = 0.5,
              color = "blue",
              na.rm = TRUE
            ) +
            ggplot2::guides(alpha = "none") +
            ggplot2::labs("") +
            ggplot2::xlab(x_label) +
            ggplot2::ylab(y_label) +
            nature_theme(input_df[, "x"], y_label) +
            ggplot2::annotate(
              geom = "text",
              x = Inf,
              y = Inf,
              hjust = 1,
              vjust = 1,
              label = sprintf(
                "FDR: %s\nCoefficient: %s\nN: %s",
                formatC(qval, format = "e", digits = 3),
                formatC(coef_val, format = "e", digits = 2),
                formatC(length(input_df[, "x"]))
              ),
              color = "black",
              size = 2,
              fontface = "italic"
            )
        } else {
          # if Metadata is categorical generate a boxplot
          ### check if the variable is categorical

          logging::loginfo(
            "Creating boxplot for categorical data, %s vs %s",
            x_label,
            y_label
          )
          input_df["x"] <- lapply(input_df["x"], as.character)

          # count the Ns for each group
          x_axis_label_names <- unique(input_df[["x"]])
          renamed_levels <- as.character(levels(metadata[, x_label]))
          if (length(renamed_levels) == 0) {
            renamed_levels <- x_axis_label_names
          }
          for (name in x_axis_label_names) {
            total <- length(which(input_df[["x"]] == name))
            new_n <- paste(name, " (n=", total, ")", sep = "")
            input_df[which(input_df[["x"]] == name), "x"] <- new_n
            renamed_levels <- replace(renamed_levels, renamed_levels == name, new_n)
          }
          input_df$xnames <- factor(input_df[["x"]], levels = renamed_levels)

          temp_plot <-
            ggplot2::ggplot(
              data = input_df, ggplot2::aes(input_df$xnames, y)
            ) +
            ggplot2::geom_boxplot(
              ggplot2::aes(fill = x),
              outlier.alpha = 0.0,
              na.rm = TRUE,
              alpha = .5,
              show.legend = FALSE
            ) +
            ggplot2::geom_point(
              ggplot2::aes(fill = x),
              alpha = 0.75,
              size = 1,
              shape = 21,
              stroke = 0.15,
              color = "black",
              position = ggplot2::position_jitterdodge()
            ) +
            ggplot2::scale_fill_brewer(palette = "Spectral")

          # format the figure to default nature format
          # remove legend, add x/y labels
          temp_plot <- temp_plot +
            nature_theme(input_df[, "x"], y_label) +
            ggplot2::theme(
              panel.grid.major = ggplot2::element_blank(),
              panel.grid.minor = ggplot2::element_blank(),
              panel.background = ggplot2::element_blank(),
              axis.line = ggplot2::element_line(colour = "black")
            ) +
            ggplot2::xlab(x_label) +
            ggplot2::ylab(y_label) +
            ggplot2::theme(legend.position = "none") +
            ggplot2::annotate(
              geom = "text",
              x = Inf,
              y = Inf,
              hjust = 1,
              vjust = 1,
              label = sprintf(
                "FDR: %s\nCoefficient: %s\nValue: %s",
                formatC(qval, format = "e", digits = 3),
                formatC(coef_val, format = "e", digits = 2),
                results_value
              ),
              color = "black",
              size = 2,
              fontface = "italic"
            )
        }
        stdout <- capture.output(print(temp_plot), type = "message")
        if (length(stdout) > 0) {
          logging::logdebug(stdout)
        }

        # keep all plots if desired
        # or only keep plots to be printed to png
        if (save_scatter) {
          saved_plots[[label]][[count]] <- temp_plot
        } else if (count <= max_pngs) {
          saved_plots[[label]][[count]] <- temp_plot
        }
        count <- count + 1
      }
      dev.off()

      # print the saved figures
      # this is done separately from pdf generation
      # because nested graphics devices cause problems in rmarkdown output
      for (plot_number in seq(1, min((count - 1), max_pngs))) {
        png_file <- file.path(
          figures_folder,
          paste0(
            substr(basename(plot_file), 1, nchar(basename(plot_file)) - 4),
            "_", plot_number, ".png"
          )
        )
        png(png_file, res = 300, width = 960, height = 960)
        stdout <- capture.output(print(saved_plots[[label]][[plot_number]]))
        dev.off()
      }
      # give plots informative names
      if (save_scatter) {
        names(saved_plots[[label]]) <- make.names(output_df_all[data_index, "feature"], unique = TRUE)
      } else {
        saved_plots[[label]] <- NULL # instead remove plots if only saved for png generation
      }
      metadata_number <- metadata_number + 1
    }
    return(saved_plots)
  }
